cgsmiles/resolve.py

```diff
@@ -185,6 +184,51 @@ def edges_from_bonding_descrpt(self):
                 order = 1
             self.molecule.add_edge(edge[0], edge[1], bonding=bonding, order=order)
 
+    def squash_atoms(self):
+        """
+        Applies the squash operator.
+        """
+        bondings = nx.get_edge_attributes(self.molecule, 'bonding')
+        squashed = False
+        for edge, bonding in bondings.items():
+            # we have a squash operator
+            if bonding[0].startswith('!'):
+                # find all hydrogens to remove
+                # and which atoms to connect
+                nodes_to_remove = [edge[1]]
+                new_edge_nodes = []
+                for hnode in self.molecule.neighbors(edge[1]):
+                    if self.molecule.nodes[hnode].get('element', 'Nan') == 'H':
+                        nodes_to_remove.append(hnode)
+                    elif hnode != edge[0]:
+                        new_edge_nodes.append(hnode)
+
+                # remove edges
+                self.molecule.remove_edge(*edge)
+                for node in nodes_to_remove[1:]:
+                    self.molecule.remove_edge(edge[1], node)
+
+                # add edges
+                for node in new_edge_nodes:
+                    self.molecule.add_edge(edge[0], node)
+
+                # find the reference hydrogen atoms
+                nodes_to_keep = [edge[0]]
+                for hnode in self.molecule.neighbors(edge[0]):
+                    if self.molecule.nodes[hnode].get('element', 'Nan') == 'H':
+                        nodes_to_keep.append(hnode)
```

Comment:
Can you explain what's going on with these hydrogen atoms?

Reply:
yes so basically when you have a squash operator like `{#A=CC[!],#B=[!]COH}` then the fragments read parsed and will result into CH3-CH2 and CH2-OH as fragments because hydrogens are added into fragments. Now when you squash you not only want to squash the C but also the adjacent hydrogens. 

Of course, one could reconstruct the hydrogen at a later point but that requires rethinking the fragment parser. 

What's your take on it @pckroon ? I can give it a try today. This project has full proiority

Reply:
I think leaving implicit hydrogens implicit until the very last moment is the smarter thing to do. That way they're not special in any way shape or form. It does mean you'll need to do a second pass at the end to guess the valences and fill up/remove them as needed, but I think that will clean up the code.

Reply:
done! Looks much neater. I also externalized the rebuilding of hydrogen into a helper function so technically you could still rebuild them for the fragments as well and have the bonding descriptors sort of dangling.